pkg/util/scheduler/queue.go

```diff
@@ -0,0 +1,505 @@
+			}
+
+			item.runnable = tq.items[0]
+			item.dequeueRespChan = pendingReq.Value.(*dequeueRequest).respChan
+			item.tenantElem = pendingTenant
+		} else {
+			s.activeTenants.Remove(pendingTenant)
+			tq.activeElement = nil
+		}
+	}
+
+	return item
+}
+
+func (s *dispatcherState) handleEnqueueRequest(req enqueueRequest) {
+	if s.closed {
+		s.discardedRequests.WithLabelValues(req.tenantID, "queue_closed").Inc()
+		req.respChan <- ErrQueueClosed
+		return
+	}
+
+	tq, exists := s.tenantQueues[req.tenantID]
+	if !exists {
+		tq = &tenantQueue{
+			id:    req.tenantID,
+			items: make([]func(), 0, 8),
+		}
+		s.tenantQueues[req.tenantID] = tq
+	}
+
+	if s.maxSizePerTenant > 0 && len(tq.items) >= s.maxSizePerTenant {
+		s.discardedRequests.WithLabelValues(req.tenantID, "queue_full").Inc()
+		req.respChan <- ErrTenantQueueFull
+		return
+	}
+
+	tq.items = append(tq.items, req.runnable)
+	s.queueLength.WithLabelValues(req.tenantID).Set(float64(len(tq.items)))
+
+	if tq.activeElement == nil {
+		tq.activeElement = s.activeTenants.PushBack(tq)
+	}
+
+	req.respChan <- nil
+}
+
+func (s *dispatcherState) handleDequeueRequest(req dequeueRequest) {
+	if s.closed {
+		s.dispatchDequeueResponse(&req, dequeueResponse{ok: false, err: ErrQueueClosed})
+		return
+	}
+
+	if req.ctx.Err() != nil {
+		s.dispatchDequeueResponse(&req, dequeueResponse{ok: false, err: req.ctx.Err()})
+		return
+	}
+
+	s.pendingRequests.PushBack(&req)
+}
+
+func (s *dispatcherState) completeDequeue(tenantElem *list.Element) {
+	reqElem := s.pendingRequests.Front()
+	s.pendingRequests.Remove(reqElem)
+
+	tq := tenantElem.Value.(*tenantQueue)
+	tq.items = tq.items[1:]
+
+	s.queueLength.WithLabelValues(tq.id).Set(float64(len(tq.items)))
+
+	if len(tq.items) == 0 {
+		s.activeTenants.Remove(tenantElem)
+		tq.activeElement = nil
+		tq.items = nil
+		delete(s.tenantQueues, tq.id)
+	} else {
+		s.activeTenants.MoveToBack(tenantElem)
+	}
+}
+
+func (s *dispatcherState) handleCloseRequest(req closeRequest) {
+	if !s.closed {
+		s.closed = true
+
+		// Notify all pending requests that the queue is closed
+		for e := s.pendingRequests.Front(); e != nil; e = e.Next() {
+			dequeueReq := e.Value.(*dequeueRequest)
+			s.dispatchDequeueResponse(dequeueReq, dequeueResponse{ok: false, err: ErrQueueClosed})
+		}
+		s.pendingRequests.Init()
+
+		// Reset queue length metrics for all tenants before clearing
+		if s.queueLength != nil {
+			for tenantID := range s.tenantQueues {
+				s.queueLength.WithLabelValues(tenantID).Set(0)
+			}
+		}
+
+		s.tenantQueues = make(map[string]*tenantQueue)
+		s.activeTenants.Init()
+	}
+
+	select {
+	case req.respChan <- struct{}{}:
+	default:
+	}
+}
+
+func (s *dispatcherState) handleLenRequest(req lenRequest) {
+	total := 0
+	for _, tq := range s.tenantQueues {
+		total += len(tq.items)
+	}
+	req.respChan <- total
+}
+
+func (s *dispatcherState) handlePendingRequests() {
+	for reqElem := s.pendingRequests.Front(); reqElem != nil && s.activeTenants.Len() > 0; {
```

Comment:
We can take both request and tenant at the same time, and only continue if we have both:

```suggestion
	for reqElem, tenantElem := s.pendingRequests.Front(), s.activeTenants.Front(); reqElem != nil && tenantElem != nil; {
```

We can also do that for next iterations, instead of updating `reqElem` manually inside the loop:

```suggestion
	for reqElem, tenantElem := s.pendingRequests.Front(), s.activeTenants.Front(); reqElem != nil && tenantElem != nil; reqElem, tenantElem = s.pendingRequests.Front(), s.activeTenants.Front() {

```